slides/aml-13-022818-resampling-imbalanced-data/index.html

```diff
@@ -0,0 +1,774 @@
+                         X_train, y_train, cv=10, scoring='roc_auc')
+print(scores.mean())
+```
+0.918
+```python
+from sklearn.ensemble import RandomForestClassifier
+scores = cross_val_score(RandomForestClassifier(n_estimators=100,
+class_weight='balanced'),X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.917
+]
+
+???
+
+---
+class: spacious
+
+# Ensemble Resampling
+
+- Random resampling separate for each instance in
+an ensemble!
+- Paper: “Exploratory Undersampling for Class Imbalance Learning”
+- Not in sklearn (yet), not totally easy with
+imbalance-learn (but soon).
+
+???
+
+---
+
+# Quick & Dirty Easy Ensemble
+
+.smaller[
+```python
+from sklearn.base import clone
+
+def make_resampled_ensemble(estimator, n_estimators=100):
+    estimators = []
+    for i in range(n_estimators):
+        est = clone(estimator)
+        if hasattr(est, "random_state"):
+            est.random_state = i
+        pipe = make_imb_pipeline(RandomUnderSampler(random_state=i,
+                                  replacement=True),est)
+        estimators.append(("est_i".format(i), pipe))
+    return VotingClassifier(estimators, voting="soft")
+
+resampled_tree_test = make_resampled_ensemble(DecisionTreeClassifier
+                                              (max_features='auto'))
+scores = cross_val_score(resampled_tree_test, X_train, y_train, cv=10,
+                                                scoring='roc_auc')
+print(np.mean(scores))
+```
+0.960
+
+
+]
+
+???
+
+-As cheap as undersampling, but much better results than anything else! </br>
+-Didn't do anything for Logistic Regression.
+
+---
+class: center, middle
+
+## Smart resampling
+## (based on nearest neighbour heuristics from the 70's)
+
+???
+
+---
+class: spacious
+
+# Edited Nearest Neighbours
+
+- Originally as heuristic for reducing dataset for KNN
+- Remove all samples that are misclassified by KNN
+from training data (mode) or that have any point
+from other class as neighbor (all).
+- “Cleans up” outliers and boundaries.
+
+???
+
+---
+
+.center[
+![:scale 90%](images/edited_nearest_neighbour.png)
+]
+
+???
+
+---
+
+# Edited Nearest Neighbours
+
+.smaller[
+```python
+from imblearn.under_sampling import EditedNearestNeighbours
+enn = EditedNearestNeighbours(n_neighbors=5)
+X_train_enn, y_train_enn = enn.fit_sample(X_train, y_train)
+
+enn_mode = EditedNearestNeighbours(kind_sel="mode", n_neighbors=5)
+X_train_enn_mode, y_train_enn_mode = enn_mode.fit_sample(X_train, y_train)
+```
+.center[
+![:scale 100%](images/edited_nearest_neighbour_2.png)
+]
+]
+
+???
+
+---
+
+.smaller[
+```python
+enn_pipe = make_imb_pipeline(EditedNearestNeighbours(n_neighbors=5),
+                             LogisticRegression())
+scores = cross_val_score(enn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.920
+```python
+enn_pipe_rf = make_imb_pipeline(EditedNearestNeighbours(n_neighbors
```

Comment:
this is a horrible line break ;)

Reply:
My bad... will correct it. I was playing with the .smaller[] too, so must have missed it